# Plink tutorial

In this tutorial, we will consider using PLINK to analyse example data: randomly selected genotypes (approximately 80,000 autosomal SNPs) from the 89 Asian HapMap individuals. A phenotype has been simulated based on the genotype at one SNP. In this tutorial, we will walk through using PLINK to work with the data, using a range of features: data management, summary statistics, population stratification and basic association analysis. 

## 89 HapMap samples and 80K random SNPs

In [3]:
# Current working directory
pwd

/Users/duz/Desktop/Center_for_Statistical_Genetics/family-association/analysis/hapmap1


In [4]:
# Change working directory
cd ./hapmap1

bash: cd: ./hapmap1: No such file or directory


: 1

The --file option takes a single parameter, the root of the input file names, and will look for two files: a PED file and a MAP file with this root name. In otherwords, --file hapmap1 implies hapmap1.ped and hapmap1.map should exist in the current directory. 

In [5]:
plink --file hapmap1

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to plink.log.
Options in effect:
  --file hapmap1

8192 MB RAM detected; reserving 4096 MB for main workspace.
.ped scan complete (for binary autoconversion).324252628293031323334353738394041424344464748495051525355565758596061626465666768697071737475767778798082838485868788899192939495969798100%
Performing single-pass .bed write (83534 variants, 89 people).
--file: plink.bed + plink.bim + plink.fam written.353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394%


### Making a binary PED file

The first thing we will do is to make a **binary PED file**. This more compact representation of the data saves space and speeds up subsequent analysis. To make a binary PED file, use the following command:

In [15]:
plink --file hapmap1 --make-bed --out hapmap1 

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to hapmap1.log.
Options in effect:
  --file hapmap1
  --make-bed
  --out hapmap1

8192 MB RAM detected; reserving 4096 MB for main workspace.
.ped scan complete (for binary autoconversion).324252628293031323334353738394041424344464748495051525355565758596061626465666768697071737475767778798082838485868788899192939495969798100%
Performing single-pass .bed write (83534 variants, 89 people).
--file: hapmap1-temporary.bed + hapmap1-temporary.bim + hapmap1-temporary.fam849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394%
written.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating

Notes:  
Three files are created with this command -- the binary file that contains the raw genotype data `hapmap1.bed` but also a revsied map file `hapmap1.bim` which contains two extra columns that give the allele names for each SNP, and `hapmap1.fam` which is just the first six columns of hapmap1.ped. You can view the .bim and .fam files -- but do not try to view the .bed file. None of these three files should be manually editted.

If, for example, you wanted to create a new file that only includes individuals with high genotyping (at least 95% complete), you would run: 

In [16]:
plink --file hapmap1 --make-bed --mind 0.05 --out highgeno 

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to highgeno.log.
Options in effect:
  --file hapmap1
  --make-bed
  --mind 0.05
  --out highgeno

8192 MB RAM detected; reserving 4096 MB for main workspace.
.ped scan complete (for binary autoconversion).324252628293031323334353738394041424344464748495051525355565758596061626465666768697071737475767778798082838485868788899192939495969798100%
Performing single-pass .bed write (83534 variants, 89 people).
--file: highgeno-temporary.bed + highgeno-temporary.bim +83940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394%
highgeno-temporary.fam written.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
0 people removed due to missing genotype data (--mind).
Using 1 thread (no multithreaded calcula

### Working with the binary PED file

To specify that the input data are in binary format, as opposed to the normal text PED/MAP format, just use the `--bfile` option instead of --file. To repeat the first command we ran (which just loads the data and prints some basic summary statistics): 

In [17]:
plink --bfile hapmap1

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to plink.log.
Options in effect:
  --bfile hapmap1


  plink <input flag(s)...> [command flag(s)...] [other flag(s)...]
  plink --help [flag name(s)...]

Commands include --make-bed, --recode, --flip-scan, --merge-list,
--write-snplist, --list-duplicate-vars, --freqx, --missing, --test-mishap,
--hardy, --mendel, --ibc, --impute-sex, --indep-pairphase, --r2, --show-tags,
--blocks, --distance, --genome, --homozyg, --make-rel, --make-grm-gz,
--rel-cutoff, --cluster, --pca, --neighbour, --ibs-test, --regress-distance,
--model, --bd, --gxe, --logistic, --dosage, --lasso, --test-missing,
--make-perm-pheno, --unrelated-heritability, --tdt, --dfam, --qfam, --tucc,
--annotate, --clump, --gene-report, --meta-analysis, --epistasis,
--fast-epistasis, and --score.

"plink --help | more" describes all functions (warning: long).


: 10

### Summary statistics: missing rates

In [10]:
plink --bfile hapmap1 --missing --out miss_stat

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to miss_stat.log.
Options in effect:
  --bfile hapmap1
  --missing
  --out miss_stat

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
--missing: Sample missing data report written to miss_stat.imiss, and
variant-based missing data report written to miss_stat.lmiss.


Notes: Here we see that no individuals were removed for low genotypes (MIND > 0.1 implies that we accept people with less than 10 percent missingness)

In [ ]:
more miss_stat.lmiss 

 CHR         SNP   N_MISS   N_GENO   F_MISS
   1   rs6681049        0       89        0
   1   rs4074137        0       89        0
   1   rs7540009        0       89        0
   1   rs1891905        0       89        0
   1   rs9729550        0       89        0
   1   rs3813196        0       89        0
   1   rs6704013        2       89  0.02247
   1    rs307347       12       89   0.1348
   1   rs9439440        2       89  0.02247
   1   rs3128342        1       89  0.01124
   1  rs12044597        0       89        0
   1  rs10907185        0       89        0
   1  rs11260616        1       89  0.01124
   1    rs745910        2       89  0.02247
   1   rs2803291       12       89   0.1348
   1   rs7531342       12       89   0.1348
   1    rs262688        0       89        0
   1   rs2460000        0       89        0
   1    rs260509        0       89        0
   1   rs2645091        0       89        0
   1   rs2643895        0       89        0
   1   rs2840529        0       

Notes: That is, for each SNP, we see the number of missing individuals (`N_MISS`) and the proportion of individuals missing (`F_MISS`).

In [5]:
more miss_stat.imiss

     FID  IID MISS_PHENO   N_MISS   N_GENO   F_MISS
  HCB181    1          N      671    83534 0.008033
  HCB182    1          N     1156    83534  0.01384
  HCB183    1          N      498    83534 0.005962
  HCB184    1          N      412    83534 0.004932
  HCB185    1          N      329    83534 0.003939
  HCB186    1          N     1233    83534  0.01476
  HCB187    1          N      258    83534 0.003089
  HCB188    1          N      864    83534  0.01034
  HCB189    1          N      517    83534 0.006189
  HCB190    1          N      519    83534 0.006213
  HCB191    1          N      303    83534 0.003627
  HCB192    1          N      319    83534 0.003819
  HCB193    1          N      401    83534   0.0048
  HCB194    1          N      411    83534  0.00492
  HCB195    1          N      667    83534 0.007985
  HCB196    1          N      308    83534 0.003687
  HCB197    1          N      271    83534 0.003244
  HCB198    1          N      506    83534 0.006057
  HCB199    

Notes: The final column is the actual genotyping rate for that individual -- we see the genotyping rate is very high here

Use PLINK to analyse the data by chromosome, with the `--chr` option. For example, to perform the above analysis for chromosome 1: 

In [6]:
plink --bfile hapmap1 --chr 1 --out res1 --missing 

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to res1.log.
Options in effect:
  --bfile hapmap1
  --chr 1
  --missing
  --out res1

8192 MB RAM detected; reserving 4096 MB for main workspace.
6762 out of 83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.995545.
--missing: Sample missing data report written to res1.imiss, and variant-based
missing data report written to res1.lmiss.


### Summary statistics: allele frequencies

Next we perform a similar analysis, except **requesting allele frequencies** instead of genotyping rates. The following command generates a file called `freq_stat.frq` which contains the minor allele frequency and allele codes for each SNP. 

In [7]:
 plink --bfile hapmap1 --freq --out freq_stat 

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to freq_stat.log.
Options in effect:
  --bfile hapmap1
  --freq
  --out freq_stat

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
--freq: Allele frequencies (founders only) written to freq_stat.frq .


In [12]:
head freq_stat.frq

 CHR         SNP   A1   A2          MAF  NCHROBS
   1   rs6681049    1    2       0.2135      178
   1   rs4074137    1    2      0.07865      178
   1   rs7540009    0    2            0      178
   1   rs1891905    1    2       0.4045      178
   1   rs9729550    1    2       0.1292      178
   1   rs3813196    1    2      0.02809      178
   1   rs6704013    0    2            0      174
   1    rs307347    0    2            0      154
   1   rs9439440    0    2            0      174


It is also possible to perform this frequency analysis (and the missingness analysis) stratified by a categorical, cluster variable. In this case, we shall use the file that indicates whether the individual is from the Chinese or the Japanese sample, pop.phe.  
To perform a stratified analysis, use the `--within` option

In [13]:
plink --bfile hapmap1 --freq --within pop.phe --out freq_stat

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to freq_stat.log.
Options in effect:
  --bfile hapmap1
  --freq
  --out freq_stat
  --within pop.phe

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
--within: 2 clusters loaded, covering a total of 89 people.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
--freq: Cluster-stratified allele frequencies (founders only) written to
freq_stat.frq.strat .


Notes: The output will now indicate that a file called `freq_stat.frq.strat`. has been generated instead of `freq_stat.frq`.  
We see each row is now the allele frequency for each SNP stratifed by subpopulation: 

In [16]:
head freq_stat.frq.strat

 CHR         SNP     CLST   A1   A2      MAF    MAC  NCHROBS
   1   rs6681049        1    1    2   0.2333     21       90 
   1   rs6681049        2    1    2   0.1932     17       88 
   1   rs4074137        1    1    2      0.1      9       90 
   1   rs4074137        2    1    2  0.05682      5       88 
   1   rs7540009        1    0    2        0      0       90 
   1   rs7540009        2    0    2        0      0       88 
   1   rs1891905        1    1    2   0.4111     37       90 
   1   rs1891905        2    1    2   0.3977     35       88 
   1   rs9729550        1    1    2   0.1444     13       90 


### Basic association analysis

In [25]:
plink --bfile hapmap1 --assoc --out as1

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to as1.log.
Options in effect:
  --assoc
  --bfile hapmap1
  --out as1

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
83534 variants and 89 people pass filters and QC.
Among remaining phenotypes, 44 are cases and 45 are controls.
Writing C/C --assoc report to as1.assoc ... 101112141516171819202223252627282931323334

In [26]:
more as1.assoc | head

 CHR         SNP         BP   A1      F_A      F_U   A2        CHISQ            P           OR 
   1   rs6681049          1    1   0.1591   0.2667    2        3.067      0.07991       0.5203 
   1   rs4074137          2    1  0.07955  0.07778    2     0.001919       0.9651        1.025 
   1   rs7540009          3    0        0        0    2           NA           NA           NA 
   1   rs1891905          4    1   0.4091      0.4    2      0.01527       0.9017        1.038 
   1   rs9729550          5    1   0.1705  0.08889    2        2.631       0.1048        2.106 
   1   rs3813196          6    1  0.03409  0.02222    2       0.2296       0.6318        1.553 
   1   rs6704013          7    0        0        0    2           NA           NA           NA 
   1    rs307347          8    0        0        0    2           NA           NA           NA 
   1   rs9439440          9    0        0        0    2           NA           NA           NA 


where each row is a single SNP association result. The fields are:

    Chromosome
    SNP identifier
    Code for allele 1 (the minor, rare allele based on the entire sample frequencies)
    The frequency of this variant in cases
    The frequency of this variant in controls
    Code for the other allele
    The chi-squared statistic for this test (1 df)
    The asymptotic significance value for this test
    The odds ratio for this test 

Sort the list of association statistics and print out the top ten:

In [27]:
sort --key=7 -nr as1.assoc | head

  22  rs13055462      82998    1   0.4318   0.2889    2        3.947      0.04695        1.871 
  22  rs13054355      82194    0        0        0    2           NA           NA           NA 
  22  rs12628795      82095    1   0.1591   0.2333    2        1.553       0.2128       0.6216 
  22  rs12628695      83122    1   0.1591   0.1556    2     0.004195       0.9484        1.027 
  22  rs12628369      83379    1   0.4545   0.4667    2      0.02631       0.8711       0.9524 
  22  rs12627919      82186    1   0.3295   0.2111    2        3.169      0.07505        1.837 
  22  rs12485231      82613    1   0.2614   0.3778    2         2.77      0.09604       0.5828 
  22  rs12484228      83136    1  0.06818  0.04444    2       0.4728       0.4917        1.573 
  22  rs12484213      82478    1  0.06818  0.07778    2      0.06052       0.8057       0.8676 
  22  rs12483834      83241    1   0.2045   0.1333    2         1.61       0.2045        1.671 


To get a sorted list of association results, that also includes a range of significance values that are **adjusted for multiple testing**, use the `--adjust` flag: 

In [28]:
plink --bfile hapmap1 --assoc --adjust --out as2

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to as2.log.
Options in effect:
  --adjust
  --assoc
  --bfile hapmap1
  --out as2

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
83534 variants and 89 people pass filters and QC.
Among remaining phenotypes, 44 are cases and 45 are controls.
Writing C/C --assoc report to as2.assoc ... 1011121415161718192022232526272

This generates the file `as2.assoc.adjust` in addition to the basic `as2.assoc` output file. Using more, one can easily look at one's most significant associations: 

In [29]:
more as2.assoc.adjusted | head

 CHR         SNP      UNADJ         GC       BONF       HOLM   SIDAK_SS   SIDAK_SD     FDR_BH     FDR_BY
  13   rs9585021  5.586e-06  4.994e-05     0.3839     0.3839     0.3188     0.3188    0.09719          1 
   2   rs2222162  5.918e-06  5.232e-05     0.4068     0.4067     0.3342     0.3342    0.09719          1 
   9  rs10810856  7.723e-06  6.483e-05     0.5308     0.5308     0.4118     0.4118    0.09719          1 
   2   rs4675607   8.05e-06  6.703e-05     0.5533     0.5533     0.4249     0.4249    0.09719          1 
   2   rs1375352  8.485e-06  6.994e-05     0.5832     0.5831     0.4419     0.4419    0.09719          1 
   2   rs4673349  8.485e-06  6.994e-05     0.5832     0.5831     0.4419     0.4419    0.09719          1 
  21    rs219746  1.228e-05  9.422e-05     0.8442     0.8441     0.5701     0.5701     0.1206          1 
   1   rs4078404  2.667e-05   0.000176          1          1     0.8401       0.84     0.2291          1 
  14   rs1152431  3.862e-05  0.0002374         

Here we see a pre-sorted list of association results. The fields are as follows:

    Chromosome
    SNP identifier
    Unadjusted, asymptotic significance value
    Genomic control adjusted significance value. This is based on a simple estimation of the inflation factor based on median chi-square statistic. These values do not control for multiple testing therefore.
    Bonferroni adjusted significance value
    Holm step-down adjusted significance value
    Sidak single-step adjusted significance value
    Sidak step-down adjusted significance value
    Benjamini & Hochberg (1995) step-up FDR control
    Benjamini & Yekutieli (2001) step-up FDR control 

In [35]:
cat as2.log

PLINK v1.90p 64-bit (16 Jun 2020)
Options in effect:
  --adjust
  --assoc
  --bfile hapmap1
  --out as2

Hostname: duzhis-mbp.dyn.nyp.org
Working directory: /Users/duz/Desktop/Center_for_Statistical_Genetics/family-association/analysis/hapmap1
Start time: Tue Sep 15 17:44:21 2020

Random number seed: 1600206261
8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values loaded from .fam.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.99441.
83534 variants and 89 people pass filters and QC.
Among remaining phenotypes, 44 are cases and 45 are controls.
Writing C/C --assoc report to as2.assoc ... done.
--adjust: Genomic inflation est. lambda (based on median chisq) = 1.25377.
--adjust values (68727 variants) written to as2.assoc.adjusted .



Notes: Genomic inflation est. lambda (based on median chisq) = 1.25377

In [36]:
plink --bfile hapmap1 --pheno pop.phe --assoc --adjust --out as3 

PLINK v1.90p 64-bit (16 Jun 2020)              www.cog-genomics.org/plink/1.9/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to as3.log.
Options in effect:
  --adjust
  --assoc
  --bfile hapmap1
  --out as3
  --pheno pop.phe

8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values present after --pheno.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.
Total genotyping rate is 0.99441.
83534 variants and 89 people pass filters and QC.
Among remaining phenotypes, 44 are cases and 45 are controls.
Writing C/C --assoc report to as3.assoc ... 10111214

In [40]:
cat as3.log

PLINK v1.90p 64-bit (16 Jun 2020)
Options in effect:
  --adjust
  --assoc
  --bfile hapmap1
  --out as3
  --pheno pop.phe

Hostname: duzhis-mbp.dyn.nyp.org
Working directory: /Users/duz/Desktop/Center_for_Statistical_Genetics/family-association/analysis/hapmap1
Start time: Tue Sep 15 17:48:07 2020

Random number seed: 1600206487
8192 MB RAM detected; reserving 4096 MB for main workspace.
83534 variants loaded from .bim file.
89 people (89 males, 0 females) loaded from .fam.
89 phenotype values present after --pheno.
Using 1 thread (no multithreaded calculations invoked).
Before main variant filters, 89 founders and 0 nonfounders present.
Calculating allele frequencies... done.
Total genotyping rate is 0.99441.
83534 variants and 89 people pass filters and QC.
Among remaining phenotypes, 44 are cases and 45 are controls.
Writing C/C --assoc report to as3.assoc ... done.
--adjust: Genomic inflation est. lambda (based on median chisq) = 1.78854.
--adjust values (68727 variants) written to

Notes: Genomic inflation est. lambda (based on median chisq) = 1.78854